pygls/workspace.py

```diff
@@ -80,14 +84,42 @@ def position_from_utf16(lines: List[str], position: Position) -> Position:
     Returns:
         The position with `character` being converted to utf-32 code units.
     """
-    try:
-        return Position(
-            line=position.line,
-            character=position.character
-            - utf16_unit_offset(lines[position.line][:position.character])
+    if len(lines) == 0:
+        return position
+    if position.line <= len(lines) - 1:
+        line = lines[position.line]
+    else:
+        log.warning(
+            f"position_from_utf16() received out of bounds index: lines[{position.line}]"
         )
-    except IndexError:
-        return Position(line=len(lines), character=0)
+        line = lines[-1]
```

Comment:
There is no need to issue a waring.
_apply_incremental_change is designed to work correctly with positions beyond the end of the file.

I  would just replace all the above with:
```python
    if position.line >= len(lines):
        return position

    line = lines[position.line]
```